# Statistical Methods for Omics Assisted Breeding
Slides for short course on 'omics assisted breeding methods.

**Dates and Location**
  Monday, November 12 to Thursday, November 15

**Lecturers**
  - Malachy Campbell (MC)
  - Hiroyoshi Iwata (HI)
  - Diego Jarquin (DJ)
  - Gota Morota (GM)
  - Emi Tanaka (ET)
  - Jessica Tressou (JT)

**What do workshop participants expect?**


**What workshop participants need**
  1. Their own laptop
  2. R v3.5 and RStudio XX installed
  3. Special packages need in the WS (BGLR, MTM, rrBLUP)
  4. For the R-Inla introduction: INLA, sp, fields, geoR, viridisLite + tidyverse

**Participants**
  Approximately 90 participants (50 - 70)

**Day 1**
  - 9:00-10:10: Basic statistical computing in R (MC) - R Markdown
     - Intro [[PDF Slides]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/Intro.pdf)
     - Part 1 [[PDF Slides]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/IntroToR1.pdf)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/IntroToR1.Rmd)
     - Part 2 [[PDF Slides]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/IntroToR2.pdf)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/IntroToR2.Rmd)
     - Part 3 [[HTML]](https://htmlpreview.github.io/?https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/IntroToR3.html)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/IntroToR3.Rmd)
     - Data [[SillyData]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/SillyData.csv)[[SeriousData]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/SeriousData.csv)[[PSA]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/IntroToR/PSA.cleaned.csv)
  - 10:10-10:20: Break
  - 10:20-12:00: Data visualization in R (ET) - ggplot2 + desplot
  - 12:00-13:00: Lunch 
  - 13:00-14:00 Introduction to Quantitative Genetics (GM)
  - 14:00-14:10: Break
  - 14:10-15:40: Least squares and Linear mixed model (MC, JT)
    - Lab slides [[PDF Slides]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/OLS.MM/OLS_MM_LabSlides.pdf)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/OLS.MM/OLS_MM_LabSlides.Rmd)
    - Exercises [[HTML]](https://htmlpreview.github.io/?https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/OLS.MM/OLS_MM_Ex.html)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/OLS.MM/OLS_MM_Ex.Rmd)
    - Data [[.csv]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/OLS.MM/MaizeRILs.csv)
  - 15:40-15:50: Break
  - 15:50-16:50 Bayesian methods (JT)
  - 17:30-19:30 Social mixer 


**Day 2**
  - 9:00-9:30 Genomic heritability (MC, GM) [[PDF Slides]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/GenomicHeritability/GenomicHerit_10-31.pdf)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/GenomicHeritability/GenomicHerit_10-31.Rmd)
  - 9:30-9:40 Break
  - 9:40-12:00 GWAS (HI)
  - 12:00-13:00 Lunch
  - 13:00-14:00 GBLUP and RR-BLUP (MC, GM)
    - Lab slides [[PDF Slides]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/tree/master/gBLUP_rrBLUP/gBLUP_RRBLUP_LabSlides.pdf)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/tree/master/gBLUP_rrBLUP/gBLUP_RRBLUP_LabSlides.Rmd)[[Data]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/tree/master/gBLUP_rrBLUP/SpindellData.zip)
    - Exercises [[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/tree/master/gBLUP_rrBLUP/gBLUPrrBLUP_LabEx.Rmd)[[HTML]](https://htmlpreview.github.io/?https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/tree/master/gBLUP_rrBLUP/gBLUPrrBLUP_LabEx.html)[[Data]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/tree/master/gBLUP_rrBLUP/ZhaoData.zip)
  - 14:00-15:00 Bayesian alphabet (DJ)
  - 15:00-15:10 Break
  - 15:10-16:10 Classical GxE including FW, AMMI, and biplot (HI, DJ)
  - 16:10-17:10 GxE covariates (HI, DJ)

**Day 3**
  - 9:00-11:00 Introduction to ASReml-R (ET)
  - 11:00-11:10 Break
  - 11:10-12:00 Theoretical part of factor analytic model (ET)
  - 12:00-13:00 Lunch
  - 13:00-14:00 Application of factor analytic model to GxE analysis with ASReml (ET)
  - 14:00-14:10 Break
  - 14:10-15:10: Bayesian factor analytic model (DJ, GM)
  - 15:10-15:20: Break
  - 15:20-16:20 Spatial analysis (ET) Gilmour et al. (1997)
  - 16:20-17:00 Introduction to R-INLA (JT)

**Day 4**
  - 9:00-12:00 Multi-trait methods for GWAS and GP (HI, DJ, GM)
  - 12:00-13:00 Lunch 
  - 13:00-15:00 Bayesian network (MC, GM) 
      - Slides[[PDF Slides]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/BN/BN_lec.pdf)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/BN/BN_lec.Rmd)
      - Exercises[[HTML]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/BN/BN_ex.html)[[.Rmd]](https://github.com/malachycampbell/StatisticalMethodsforOmicsAssistedBreeding/blob/master/BN/BN_ex.Rmd)
  - 15:00-15:10 Break
  - 15:10-17:00 Experimental design + Optimal designs (ET)


**Useful software and other resources:**
- [CNS genomics](http://cnsgenomics.com/software.html): This is Peter Visscher's group. Really, too many useful resources to list. 

- [TASSEL](http://www.maizegenetics.net/tassel): This is a great software that can be used for GWAS and genomic prediction. It is available as a standalone GUI or can be executed from the command line. It can read in SNP data in Hapmap, HDF5 (Hierarchical Data Format version 5), VCF (Variant Call Format), Plink, Projection Alignment, Phylip, FASTA, and Numerical formats. Includes pipelines for imputation and GBS. Conveniently, there is a huge community that have provided great documentation and tutorials, and have an active message board can help with issues.

- [GAPIT](http://www.zzlab.net/GAPIT/): A nice R package that can be used for GWAS and genomic prediction. Pretty straight forward to use.

- [Jeff Endelman's Page](https://potatobreeding.cals.wisc.edu/software/): The author of the rrBLUP package.

- [PBGworks YouTube channel](https://www.youtube.com/plantbreedgenomics): Some nice tutorials for GWAS and genomic prediction.
    - Also check out their [extension page](https://articles.extension.org/plant_breeding_genomics).
  
- [Bayesian Generalized Linear Regression (BGLR) package](https://cran.r-project.org/web/packages/BGLR/index.html): An R packages that performs several types of Bayesian regression models. Check out their original [publication](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4196607/) and [tutorial](http://bglr.r-forge.r-project.org/BGLR-tutorial.pdf) for more information.

- [Rice Diversity](http://ricediversity.org/): If you need some data, this is a great place to access good phenotypic and genotypic data. 
